File: servers/go/internal/cidi/signature.go
package cidi

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"sort"
	"strconv"
	"strings"
	"time"
)

type SignableParams map[string]any

func CreateNonce() string {
	bytes := make([]byte, 16)
	if _, err := rand.Read(bytes); err != nil {
		return strconv.FormatInt(time.Now().UnixNano(), 10)
	}
	return hex.EncodeToString(bytes)
}

func NowInSeconds() int64 {
	return time.Now().Unix()
}

func BuildSignString(params SignableParams, timestamp string, nonce string) string {
	keys := make([]string, 0, len(params))
	for key, value := range params {
		if shouldSignValue(value) {
			keys = append(keys, key)
		}
	}
	sort.Strings(keys)

	parts := make([]string, 0, len(keys)+2)
	for _, key := range keys {
		parts = append(parts, key+"="+StringifySignValue(params[key]))
	}

	parts = append(parts, "timestamp="+timestamp)
	parts = append(parts, "nonce="+nonce)

	return strings.Join(parts, "&")
}

func GenerateSignature(params SignableParams, timestamp string, nonce string, secret string) string {
	signString := BuildSignString(params, timestamp, nonce)
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(signString))
	return hex.EncodeToString(mac.Sum(nil))
}

func VerifySignature(params SignableParams, timestamp string, nonce string, signature string, secret string) bool {
	expected := GenerateSignature(params, timestamp, nonce, secret)
	return hmac.Equal([]byte(expected), []byte(signature))
}

func RemoveEmptyValues(params SignableParams) SignableParams {
	cleaned := SignableParams{}
	for key, value := range params {
		if shouldSignValue(value) {
			cleaned[key] = value
		}
	}
	return cleaned
}

func StringifySignValue(value any) string {
	switch typed := value.(type) {
	case string:
		return typed
	case bool:
		return strconv.FormatBool(typed)
	case int:
		return strconv.Itoa(typed)
	case int64:
		return strconv.FormatInt(typed, 10)
	case float64:
		return strconv.FormatFloat(typed, 'f', -1, 64)
	case float32:
		return strconv.FormatFloat(float64(typed), 'f', -1, 32)
	case json.Number:
		return typed.String()
	default:
		bytes, err := json.Marshal(typed)
		if err == nil {
			return string(bytes)
		}
		return fmt.Sprint(typed)
	}
}

func shouldSignValue(value any) bool {
	if value == nil {
		return false
	}
	return fmt.Sprint(value) != ""
}

File: servers/python/src/cidi/signature.py
import hashlib
import hmac
import json
import secrets
import time
from typing import Any, Mapping


SignableParams = Mapping[str, Any]


def create_nonce() -> str:
    return secrets.token_hex(16)


def now_in_seconds() -> int:
    return int(time.time())


def build_sign_string(params: SignableParams, timestamp: int | str, nonce: str) -> str:
    parts = []

    for key in sorted(params.keys()):
        value = params[key]
        if value is None or str(value) == "":
            continue
        parts.append(f"{key}={stringify_sign_value(value)}")

    parts.append(f"timestamp={timestamp}")
    parts.append(f"nonce={nonce}")

    return "&".join(parts)


def generate_signature(params: SignableParams, timestamp: int | str, nonce: str, secret: str) -> str:
    sign_string = build_sign_string(params, timestamp, nonce)
    return hmac.new(secret.encode("utf-8"), sign_string.encode("utf-8"), hashlib.sha256).hexdigest()


def verify_signature(
    params: SignableParams,
    timestamp: int | str,
    nonce: str,
    signature: str,
    secret: str,
) -> bool:
    expected = generate_signature(params, timestamp, nonce, secret)
    return hmac.compare_digest(expected, signature)


def remove_empty_values(params: SignableParams) -> dict[str, Any]:
    return {key: value for key, value in params.items() if value is not None and str(value) != ""}


def stringify_sign_value(value: Any) -> str:
    if isinstance(value, bool):
        return "true" if value else "false"

    if isinstance(value, (dict, list)):
        return json.dumps(value, separators=(",", ":"), ensure_ascii=False)

    return str(value)
